src/Agent.Worker/TestResults/TestDataPublisher.cs

```diff
@@ -117,6 +120,108 @@ public async Task<bool> PublishAsync(TestRunContext runContext, List<string> tes
             return false;
         }
 
+        public async Task<bool> PublishAsync(TestRunContext runContext, List<string> testResultFiles, TestCaseResult[] testCaseResults, PublishOptions publishOptions, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                TestDataProvider testDataProvider = ParseTestResultsFile(runContext, testResultFiles);
+                var publishTasks = new List<Task>();
+
+                if (testDataProvider != null)
+                {
+                    var testRunData = testDataProvider.GetTestRunData();
+
+                    if (!testCaseResults.IsNullOrEmpty())
+                    {
+                        //Dictionary because FQN to Test Case Results is 1 to many
+                        Dictionary<string, List<TestCaseResult>> testResultByFQN = new();
+
+                        // Iterate through the list of objects
+                        foreach (TestCaseResult testResult in testCaseResults)
+                        {
+                            if (!testResultByFQN.ContainsKey(testResult.AutomatedTestName))
+                            {
+                                // If not, initialize the list associated with the key
+                                testResultByFQN[testResult.AutomatedTestName] = new List<TestCaseResult>();
+                            }
+                            // Add the object to the dictionary using its Id as the key
+                            testResultByFQN[testResult.AutomatedTestName].Add(testResult);
+                        }
+
+                        int testRunDataIterator = 0;
+                        int testResultDataIterator = 0;
+
+                        for (testRunDataIterator = 0; testRunDataIterator < testRunData.Count; testRunDataIterator++)
```

Comment:
How big are these sets of test run data and test results?

Reply:
We have threshold for test run data as 100, if it more than 100, we merge them and 1 test run data is formed. For Test results, it is on average around 100 and the worst seen so far is 1000.